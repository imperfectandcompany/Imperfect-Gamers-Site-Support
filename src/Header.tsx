// ~/Header.tsx

import { FunctionalComponent } from 'preact';
import { Link } from 'preact-router/match';
import { content } from './content';

interface HeaderProps {
  onSearchChange: (event: Event) => void;
  onLogoClick: () => void;
  searchQuery: string;
}

export const Header: FunctionalComponent<HeaderProps> = ({
  onSearchChange,
  onLogoClick,
  searchQuery,
}) => {
  return (
    <header className="flex justify-between items-center px-4">
      <div className="flex items-center cursor-pointer z-30">
        <Link href="/">
          <img
            alt="Imperfect Gamers Logo, a gaming community"
            src="https://cdn.imperfectgamers.org/inc/assets/img/logo.svg"
            style={{ height: "60px" }}
            onClick={onLogoClick}
          />
        </Link>
      </div>
      <div className="flex space-x-4">
        <input
          type="text"
          placeholder="Search..."
          value={searchQuery} // Controlled input value
          onInput={onSearchChange} // Changed to onInput for Preact
          className="block w-full h-10 px-4 py-3 text-indigo-300 border rounded-lg appearance-none focus:border-stone-300 focus:outline-none focus:ring-stone-300 sm:text-sm pr-14"
        />
        <button
          className="flex items-center justify-center w-auto h-10 px-4 py-2 text-sm font-semibold text-white transition-all rounded-lg hover:to-indigo-600 bg-gradient-to-b from-indigo-300 via-indigo-400 to-indigo-500 outline-none"
          aria-label="Contact us for help"
        >
          <span>{content.header.contactUs}</span>
        </button>
      </div>
    </header>
  );
};
