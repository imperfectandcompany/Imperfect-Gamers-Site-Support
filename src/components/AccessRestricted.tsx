// src/components/AccessRestricted.tsx
import { FunctionalComponent } from "preact";
import { Link } from "preact-router/match";

interface AccessRestrictedProps {
  message: string;
}

export const AccessRestricted: FunctionalComponent<AccessRestrictedProps> = ({ message }) => {
  return (
    <div className="container mx-auto relative px-8 py-16 max-w-7xl md:px-12 lg:px-18 lg:py-22">
      <h1 className="text-3xl font-normal tracking-tighter text-black sm:text-4xl lg:text-5xl">
        {message}
      </h1>
      <Link href="/" className="text-indigo-500 hover:underline mt-8 inline-block">
        Go to Home
      </Link>
    </div>
  );
};