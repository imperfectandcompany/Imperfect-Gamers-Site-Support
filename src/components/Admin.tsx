import { useState } from "preact/hooks";
import { content } from "../content";
import { FunctionalComponent } from "preact";

export const Admin: FunctionalComponent = ({ }) => {
  const [editMode, setEditMode] = useState<{ section: string, id: number } | null>(null);
  const [activeSection, setActiveSection] = useState<string | null>(null);
  const [cardStates, setCardStates] = useState(content.sections);

  const handleEdit = (section: string, id: number) => {
    setEditMode({ section, id });
  };

  const handleCloseEdit = () => {
    setEditMode(null);
  };

  const toggleSection = (sectionKey: string) => {
    setActiveSection(prevSection => (prevSection === sectionKey ? null : sectionKey));
  };

  const toggleArchive = (sectionKey: string, cardIndex: number) => {
    setCardStates(prevStates => {
      const newState = { ...prevStates };
      const section = newState[sectionKey];
      const card = section.cards[cardIndex];
      card.archived = !card.archived;
      return newState;
    });
  };

  const toggleStaffOnly = (sectionKey: string, cardIndex: number) => {
    setCardStates(prevStates => {
      const newState = { ...prevStates };
      const section = newState[sectionKey];
      const card = section.cards[cardIndex];
      card.staffOnly = !card.staffOnly;
      return newState;
    });
  };

  return (
    <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8">
      <h1 className="text-3xl font-bold text-gray-900 mt-6">Admin Dashboard</h1>
      <section className="mt-8">
        <h2 className="text-xl font-bold text-indigo-600">Content Management</h2>
        {Object.keys(cardStates).map((sectionKey) => {
          const section = cardStates[sectionKey];
          const isActive = activeSection === sectionKey;
          return (
            <div key={sectionKey} className="mt-5">
              <button
                onClick={() => toggleSection(sectionKey)}
                className="flex justify-between items-center w-full text-left text-lg font-semibold text-indigo-500 py-2 transition duration-300 ease-in-out transform hover:scale-100 focus:outline-none"
              >
                {section.title} ({section.cards.length} Articles)
                <span className={`transform transition-transform duration-300 ${isActive ? 'rotate-180' : 'rotate-0'}`}>
                  <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
                  </svg>
                </span>
              </button>
              <div className={`transition-max-height duration-500 ease-in-out overflow-hidden ${isActive ? 'max-h-screen' : 'max-h-0'}`}>
                <div className="border-b border-gray-200">
                  {section.cards.map((card, index) => (
                    <div
                      key={index}
                      className={`flex justify-between items-center mb-4 transition duration-300 ease-in-out p-4 transform ${card.archived ? 'bg-stone-200' : ''} ${card.staffOnly ? 'border-l-4 border-blue-600' : ''}`}
                    >
                      <div className='m2'>
                        <h4 className="font-medium text-lg">{card.title}</h4>
                        <p className={'mt-'}>{card.description}</p>
                        <div className="flex space-x-4 text-sm mt-1">
                          <button
                            onClick={() => toggleArchive(sectionKey, index)}
                            className="text-blue-500 hover:underline"
                          >
                            {card.archived ? 'Unarchive' : 'Archive'}
                          </button>
                          <button
                            onClick={() => toggleStaffOnly(sectionKey, index)}
                            className="text-blue-500 hover:underline"
                          >
                            {card.staffOnly ? 'Make Public' : 'Make Staff Only'}
                          </button>
                        </div>
                      </div>
                      <button
                        onClick={() => handleEdit(sectionKey, index)}
                        className="text-indigo-500 hover:text-indigo-800 transition duration-300 ease-in-out"
                      >
                        Edit
                      </button>
                    </div>
                  ))}
                </div>
              </div>
            </div>
          );
        })}
      </section>
      {editMode && (
        <div className="fixed inset-0 bg-black bg-opacity-50 flex justify-center items-center transition-opacity duration-300 ease-in-out">
          <div className="bg-white p-5 rounded-lg shadow-lg transform transition-all duration-300 ease-in-out scale-105">
            <h2 className="text-xl mb-4">Edit {editMode.section}</h2>
            <button
              onClick={handleCloseEdit}
              className="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded transition duration-300 ease-in-out"
            >
              Close
            </button>
          </div>
        </div>
      )}
    </div>
  );
};
