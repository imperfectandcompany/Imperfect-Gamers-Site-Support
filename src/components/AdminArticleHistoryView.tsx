// src/components/AdminArticleHistoryView

// src/components/AdminArticleHistoryView.tsx

import { FunctionalComponent } from "preact";
import { Card } from "../content";
import { useEffect, useState } from "preact/hooks";
import { useMockAuth } from "./models/userModel";

interface AdminArticleHistoryViewProps {
  card: Card;
}

export const AdminArticleHistoryView: FunctionalComponent<AdminArticleHistoryViewProps> = ({ card }) => {
  const [usernames, setUsernames] = useState<{ [key: number]: string }>({});
  const { getUsernameById } = useMockAuth();

  useEffect(() => {
    const fetchUsernames = async () => {
      const uniqueUserIds = [
        ...new Set(card.versions.map((version) => version.editedBy)),
      ];

      const usernamePromises = uniqueUserIds.map(async (userId) => {
        const username = await getUsernameById(userId);
        return { userId, username };
      });

      const usernameResults = await Promise.all(usernamePromises);
      const usernameMap = usernameResults.reduce((acc, { userId, username }) => {
        acc[userId] = username;
        return acc;
      }, {} as { [key: number]: string });

      setUsernames(usernameMap);
    };

    fetchUsernames();
  }, [card.versions]);

  return (
    <div>
      <h2>Change History</h2>
      {card.versions.map((version) => (
        <div key={version.versionId}>
          <p>
            Version {version.versionId} ({version.versionId === 1 ? "Created" : "Edited"} by {usernames[version.editedBy] || "Loading..."} on {new Date(version.editDate).toLocaleString()})
          </p>
          {version.changes && (
            <ul>
              {version.changes.map((change, index) => (
                <li key={index}>{change}</li>
              ))}
            </ul>
          )}
        </div>
      ))}
    </div>
  );
};
