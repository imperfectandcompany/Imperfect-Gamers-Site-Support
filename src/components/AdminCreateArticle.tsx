// src/components/AdminCreateArticle.tsx
import { FunctionalComponent, h } from "preact";
import { useState, useEffect, useRef } from "preact/hooks";
import { route } from "preact-router";
import { useMockAuth } from "./models/userModel";
import { addNewArticle, checkArticleExists, getAllCategories } from "../utils";

export const AdminCreateArticle: FunctionalComponent = () => {
  const { isAuthenticated } = useMockAuth();
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [detailedDescription, setDetailedDescription] = useState("");
  const [imgSrc, setImgSrc] = useState("");
  const [category, setCategory] = useState("");
  const [categories, setCategories] = useState<
    { key: string; title: string }[]
  >([]);
  const [loading, setLoading] = useState(false);
  const [titleExists, setTitleExists] = useState(false);
  const [failedRegex, setFailedRegex] = useState(false);
  const formRef = useRef<HTMLFormElement>(null);
  const imgDebounceRef = useRef<NodeJS.Timeout | null>(null);

  const titleDebounceRef = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    async function fetchCategories() {
      const categoriesList = await getAllCategories();
      setCategories(categoriesList);
      if (categoriesList.length > 0) {
        setCategory(categoriesList[0].key);
      }
    }
    fetchCategories();
  }, []);

  const handleTitleChange = (newTitle: string) => {
    setTitle(newTitle);
    if (titleDebounceRef.current) clearTimeout(titleDebounceRef.current);
    titleDebounceRef.current = setTimeout(async () => {
      setTitleExists(await checkArticleExists(newTitle));
    }, 500);
  };

  const handleImgChange = (imgSource: string) => {
    setImgSrc(imgSource);
    if (imgDebounceRef.current) clearTimeout(imgDebounceRef.current);
    imgDebounceRef.current = setTimeout(async () => {
      setFailedRegex(!imgSource.match(new RegExp('https?://.+\\.(jpg|jpeg|png|gif)$')));
    }, 500);
  };

  const handleCreate = async (event: Event) => {
    event.preventDefault();
    if (!formRef.current?.checkValidity() || titleExists) {
      formRef.current?.classList.add("shake");
      setTimeout(() => formRef.current?.classList.remove("shake"), 500);
      const firstInvalidElement = formRef.current?.querySelector(
        "input:invalid, textarea:invalid, select:invalid"
      ) as HTMLInputElement;
      firstInvalidElement?.focus();
      return;
    }
    setLoading(true);
    const result = await addNewArticle({
      title,
      description,
      detailedDescription,
      category,
      imgSrc,
    });
    setLoading(false);
    if (result.success) {
      alert("Article created successfully!");
      route("/admin/dashboard");
    } else {
      alert("Failed to create article");
    }
  };

  if (!isAuthenticated()) {
    route("/admin");
  }

  return (
    <div className="create-article-form relative px-8 py-16 mx-auto max-w-7xl md:px-12 lg:px-18 lg:py-22">
      <h1 className="text-3xl font-normal tracking-tighter text-black sm:text-4xl lg:text-5xl mb-8">
        Create New Article
      </h1>
      <form
        ref={formRef}
        onSubmit={handleCreate}
        noValidate
        className="space-y-4"
      >
        <InputField
          label="Title"
          type="text"
          value={title}
          onChange={handleTitleChange}
          exists={titleExists}
          errorMessage="This title already exists. Please use a different title."
        />
        <TextAreaField
          label="Description"
          value={description}
          onChange={setDescription}
        />
        <TextAreaField
          label="Detailed Description"
          value={detailedDescription}
          onChange={setDetailedDescription}
        />
        <InputField
          label="Image Source"
          type="text"
          value={imgSrc}
          onChange={handleImgChange}
          failedRegex={failedRegex}
          errorMessage="Please enter a valid image URL ending with .jpg, .jpeg, .png, or .gif"
        />
        <SelectField
          label="Category"
          options={categories}
          selectedValue={category}
          onChange={setCategory}
        />
        <button
          className="bg-indigo-700 hover:bg-indigo-800 text-white p-3 w-full mt-4 transition-all duration-300 disabled:opacity-50"
          type="submit"
          disabled={loading || titleExists || failedRegex}
        >
          {loading ? "Creating..." : "Create Article"}
        </button>
      </form>
    </div>
  );
};

const InputField = ({
  label,
  type,
  value,
  onChange,
  exists = false,
  failedRegex = false,
  pattern,
  errorMessage,
}: {
  label: string;
  type: string;
  value: string;
  onChange: (value: string) => void;
  exists?: boolean;
  failedRegex?: boolean;
  pattern?: string;
  errorMessage?: string;
}) => (
  <div className="block">
    <label>{label}:</label>
    <input
      type={type}
      className={`border p-2 w-full transition duration-200 focus:outline-none ${
        exists
          ? "border-red-500"
          : "hover:border-gray-400 focus:border-indigo-500"
      }`}
      value={value}
      onInput={(e) => onChange((e.currentTarget as HTMLInputElement).value)}
      required
      pattern={pattern}
    />
    {exists || failedRegex && <p className="text-red-500 text-sm mt-1">{errorMessage}</p>}
  </div>
);

const TextAreaField = ({
  label,
  value,
  onChange,
}: {
  label: string;
  value: string;
  onChange: (value: string) => void;
}) => (
  <label className="block">
    {label}:
    <textarea
      className="border border-gray-300 hover:border-gray-400 p-2 w-full transition duration-200 focus:outline-none"
      value={value}
      onInput={(e) => onChange(e.currentTarget.value)}
      required
    />
  </label>
);

const SelectField = ({
  label,
  options,
  selectedValue,
  onChange,
}: {
  label: string;
  options: { key: string; title: string }[];
  selectedValue: string;
  onChange: (value: string) => void;
}) => (
  <label className="block">
    {label}:
    <select
      className="border border-gray-300 hover:border-gray-400 p-2 w-full transition duration-200 focus:outline-none"
      onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
        onChange(e.currentTarget.value)
      }
      value={selectedValue}
      required
    >
      {options.map((option) => (
        <option key={option.key} value={option.key}>
          {option.title}
        </option>
      ))}
    </select>
  </label>
);
