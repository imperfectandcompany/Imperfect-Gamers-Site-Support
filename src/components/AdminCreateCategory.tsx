import { FunctionalComponent, h } from "preact";
import { useState, useRef } from "preact/hooks";
import { route } from "preact-router";
import { checkCategoryExists, addNewCategory } from "../utils";
import { useMockAuth } from "./models/userModel";

export const AdminCreateCategory: FunctionalComponent = () => {
  const [categoryName, setCategoryName] = useState("");
  const [loading, setLoading] = useState(false);
  const [categoryExists, setCategoryExists] = useState(false);
  const formRef = useRef<HTMLFormElement>(null);
  const { isAuthenticated } = useMockAuth();

  const debounceRef = useRef<NodeJS.Timeout | null>(null);

  const handleCategoryChange = (newCategory: string) => {
    setCategoryName(newCategory);
    if (debounceRef.current) clearTimeout(debounceRef.current);
    debounceRef.current = setTimeout(async () => {
      setCategoryExists(await checkCategoryExists(newCategory));
    }, 500);
  };



  if (!isAuthenticated()) {
    route("/admin");
  }


  const handleCreate = async (event: Event) => {
    event.preventDefault();
    if (formRef.current?.checkValidity() === false || categoryExists) {
      formRef.current?.classList.add("shake");
      setTimeout(() => formRef.current?.classList.remove("shake"), 500);
      return;
    }
    setLoading(true);
    const result = await addNewCategory({ title: categoryName });
    setLoading(false);
    if (result.success) {
      alert("Category created successfully!");
      route("/admin/dashboard");
    } else {
      alert("Failed to create category");
    }
  };

  return (
    <div className="px-8 py-16 mx-auto max-w-7xl md:px-12 lg:px-18 lg:py-22">
      <h1 className="text-3xl font-normal tracking-tighter text-black sm:text-4xl lg:text-5xl mb-8">
        Create New Category
      </h1>
      <form ref={formRef} onSubmit={handleCreate} noValidate className="space-y-4">
        <div className="block">
          <label className="block text-sm font-medium text-gray-700">
            Category Name:
          </label>
          <input
            type="text"
            className={`mt-1 block w-full px-3 py-2 border ${
              categoryExists ? "border-red-500" : "border-gray-300"
            } rounded-md shadow-sm focus:outline-none focus:ring-indigo-500 focus:border-indigo-500`}
            value={categoryName}
            onInput={(e) => handleCategoryChange((e.target as HTMLInputElement).value)}
            required
          />
          {categoryExists && (
            <p className="mt-2 text-sm text-red-600" id="category-error">
              Category already exists. Please use a different name.
            </p>
          )}
        </div>
        <button
          type="submit"
          disabled={loading || categoryExists}
          className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline disabled:opacity-50 transition ease-in-out duration-300"
        >
          {loading ? "Creating..." : "Create Category"}
        </button>
      </form>
    </div>
  );
};
