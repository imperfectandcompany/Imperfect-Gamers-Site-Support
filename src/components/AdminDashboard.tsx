import { useEffect, useState } from "preact/hooks";
import { content } from "../content";
import { FunctionalComponent } from "preact";
import { route } from "preact-router";
import Breadcrumb from "./Breadcrumb";
import { useMockAuth } from "./models/userModel";

export const AdminDashboard: FunctionalComponent = ({ }) => {
  const [editMode, setEditMode] = useState<{ section: string, id: number } | null>(null);
  const [activeSection, setActiveSection] = useState<string | null>(null);
  const [cardStates, setCardStates] = useState(content.sections);
  const { isAuthenticated } = useMockAuth();

  useEffect(() => {
    if (!isAuthenticated()) {
      route('/admin'); // Redirect to dashboard if already logged in
    }
  }, []);


  const handleEdit = (id: number) => {
    // Correctly pass the article ID to the AdminEditArticle component
    route(`/admin/edit/${id}`);
};

  const handleCloseEdit = () => {
    setEditMode(null);
  };

  const toggleSection = (sectionKey: string) => {
    setActiveSection(prevSection => (prevSection === sectionKey ? null : sectionKey));
  };

  const toggleArchive = (sectionKey: string, cardIndex: number) => {
    setCardStates(prevStates => {
      const newState = { ...prevStates };
      const section = newState[sectionKey];
      const card = section.cards[cardIndex];
      card.archived = !card.archived;
      return newState;
    });
  };

  const toggleStaffOnly = (sectionKey: string, cardIndex: number) => {
    setCardStates(prevStates => {
      const newState = { ...prevStates };
      const section = newState[sectionKey];
      const card = section.cards[cardIndex];
      card.staffOnly = !card.staffOnly;
      return newState;
    });
  };

  return (
    <div>
      <Breadcrumb path={`/admin`} />
      <div className="container relative px-8 py-16 mx-auto max-w-7xl md:px-12 lg:px-18 lg:py-22">
        <h1 className="text-3xl font-normal tracking-tighter text-black sm:text-4xl lg:text-5xl">
          Admin Dashboard
        </h1>

        <div className="mt-8">
  <button
    onClick={() => route('/admin/logs')}
    className="px-4 py-2 bg-indigo-500 text-white font-bold rounded hover:bg-indigo-600 transition duration-300 ease-in-out"
  >
    Visit Admin Logs
  </button>
  </div>

      <section className="mt-8">
        <h2 className="text-xl font-bold text-indigo-600">Content Management</h2>
        {Object.keys(cardStates).map((sectionKey) => {
            const section = cardStates[sectionKey];
            const latestSectionVersion = section.versions.slice(-1)[0];
            const isActive = activeSection === sectionKey;
          return (
            <div key={sectionKey} className="mt-5">
                
              <button
                onClick={() => toggleSection(sectionKey)}
                className="flex justify-between items-center w-full text-left text-lg font-semibold text-indigo-500 py-2 transition duration-300 ease-in-out transform hover:scale-100 focus:outline-none"
              >
                  {latestSectionVersion.title} ({section.cards.length} Articles)
                <span className={`transform transition-transform duration-300 ${isActive ? 'rotate-180' : 'rotate-0'}`}>
                  <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
                  </svg>
                </span>
              </button>
              <div className={`transition-max-height duration-500 ease-in-out overflow-hidden ${isActive ? 'max-h-screen' : 'max-h-0'}`}>
                <div className="border-b border-gray-200">
                {section.cards.map((card, index) => {
                      const latestCardVersion = card.versions.slice(-1)[0];
                      return (
                    <div
                      key={index}
                      className={`flex justify-between items-center mb-4 transition duration-300 ease-in-out p-4 transform ${card.archived ? 'bg-stone-50' : ''} ${card.staffOnly ? 'border-l-4 border-indigo-600' : ''}`}
                    >
                      <div className='m2'>
                      <h4 className="font-medium text-lg">{latestCardVersion.title}</h4>
                      <p className='mt-1'>{latestCardVersion.description}</p>
                        <div className="flex space-x-4 text-sm mt-1">
                          <button
                            onClick={() => toggleArchive(sectionKey, index)}
                            className="text-indigo-500 hover:underline"
                          >
                            {card.archived ? 'Unarchive' : 'Archive'}
                          </button>
                          <button
                            onClick={() => toggleStaffOnly(sectionKey, index)}
                            className="text-indigo-500 hover:underline"
                          >
                            {card.staffOnly ? 'Make Public' : 'Make Staff Only'}
                          </button>
                        </div>
                      </div>
                      <button
                        onClick={() => handleEdit(card.id)}
                        className="text-indigo-500 hover:text-indigo-800 transition duration-300 ease-in-out"
                      >
                        Edit
                      </button>
                    </div>
                      );
                    })}
                </div>
              </div>
            </div>
          );
        })}
      </section>
      {editMode && (
        <div className="fixed inset-0 bg-black bg-opacity-50 flex justify-center items-center transition-opacity duration-300 ease-in-out">
          <div className="bg-white p-5 rounded-lg shadow-lg transform transition-all duration-300 ease-in-out scale-105">
            <h2 className="text-xl mb-4">Edit {editMode.section}</h2>
            <button
              onClick={handleCloseEdit}
              className="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded transition duration-300 ease-in-out"
            >
              Close
            </button>
          </div>
        </div>
      )}
    </div>
    </div>
  );
};
