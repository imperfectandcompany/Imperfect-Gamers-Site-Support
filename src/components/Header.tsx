// src/components/Header.tsx

import { FunctionalComponent } from "preact";
import { Link } from "preact-router/match";

interface HeaderProps {
  onSearchChange: (event: Event) => void;
  onLogoClick: () => void;
  onCategoryClick: () => void;
  searchQuery: string | null;
}

export const Header: FunctionalComponent<HeaderProps> = ({
  onSearchChange,
  onLogoClick,
  onCategoryClick,
  searchQuery,
}) => {
  return (
    <header className="flex justify-between items-center px-4">
      <div className="flex items-center cursor-pointer z-30">
        <Link href="/" onClick={onLogoClick}>
          <img
            alt="Imperfect Gamers Logo, a gaming community"
            src="https://cdn.imperfectgamers.org/inc/assets/img/logo.svg"
            style={{ height: "60px" }}
          />
        </Link>
      </div>
      <div className="flex space-x-4">
        <input
          type="text"
          placeholder="Search..."
          value={searchQuery ?? ""} // Ensure value is a string
          onInput={onSearchChange} // Changed to onInput for Preact
          className="block w-full h-10 px-4 py-3 text-indigo-300 border rounded-lg appearance-none focus:border-stone-300 focus:outline-none focus:ring-stone-300 sm:text-sm pr-14"
        />
        <Link
          href="/categories"
          className="flex items-center h-10 px-4 py-3 text-indigo-300 border rounded-lg appearance-none focus:border-stone-300 focus:outline-none focus:ring-stone-300 sm:text-sm"
          onClick={onCategoryClick}
        >
          Categories
        </Link>
      </div>
    </header>
  );
};
