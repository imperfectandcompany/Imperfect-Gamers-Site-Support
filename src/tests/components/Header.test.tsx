import { render, screen, fireEvent } from '@testing-library/preact';
import { describe, it, expect, vi } from 'vitest';
import { Header } from '../../components/Header';

describe('Header', () => {
  it('renders the header with logo and search input', () => {
    render(<Header onSearchChange={() => {}} onLogoClick={() => {}} searchQuery="" />);
    
    expect(screen.getByAltText(/Imperfect Gamers Logo, a gaming community/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/Search.../i)).toBeInTheDocument();
  });

  it('calls onSearchChange when typing in search input', () => {
    const mockOnSearchChange = vi.fn();
    render(<Header onSearchChange={mockOnSearchChange} onLogoClick={() => {}} searchQuery="" />);

    const searchInput = screen.getByPlaceholderText(/Search.../i);
    fireEvent.input(searchInput, { target: { value: 'Surf Maps' } });

    expect(mockOnSearchChange).toHaveBeenCalledWith(expect.any(Object)); // Expecting the event object
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<Header onSearchChange={() => {}} onLogoClick={() => {}} searchQuery="" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onLogoClick when logo is clicked', () => {
    const mockOnLogoClick = vi.fn();
    render(<Header onSearchChange={() => {}} onLogoClick={mockOnLogoClick} searchQuery="" />);

    const logo = screen.getByAltText(/Imperfect Gamers Logo, a gaming community/i);
    fireEvent.click(logo);

    expect(mockOnLogoClick).toHaveBeenCalled();
  });
});
