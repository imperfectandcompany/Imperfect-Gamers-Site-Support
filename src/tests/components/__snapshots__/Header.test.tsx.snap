// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > matches the snapshot 1`] = `
<DocumentFragment>
  <header
    class="flex justify-between items-center px-4"
  >
    <div
      class="flex items-center cursor-pointer z-30"
    >
      <a
        class=""
        href="/"
      >
        <img
          alt="Imperfect Gamers Logo, a gaming community"
          src="https://cdn.imperfectgamers.org/inc/assets/img/logo.svg"
          style="height: 60px;"
        />
      </a>
    </div>
    <div
      class="flex space-x-4"
    >
      <input
        class="block w-full h-10 px-4 py-3 text-indigo-300 border rounded-lg appearance-none focus:border-stone-300 focus:outline-none focus:ring-stone-300 sm:text-sm pr-14"
        placeholder="Search..."
        type="text"
      />
      <a
        aria-label="Categories"
        class="flex items-center justify-center w-auto h-10 px-4 py-2 text-sm font-semibold text-white transition-all rounded-lg hover:to-indigo-600 bg-gradient-to-b from-indigo-300 via-indigo-400 to-indigo-500 outline-none"
        href="/categories"
      >
        Categories
      </a>
      <button
        aria-label="Contact us for help"
        class="flex items-center justify-center w-auto h-10 px-4 py-2 text-sm font-semibold text-white transition-all rounded-lg hover:to-indigo-600 bg-gradient-to-b from-indigo-300 via-indigo-400 to-indigo-500 outline-none"
      >
        <span>
          Contact
        </span>
      </button>
    </div>
  </header>
</DocumentFragment>
`;
